### PR TITLE
Fix #102 - Update Part2 Step17 to match new requirements

### LINKED ISSUE #102: Update Part 2 Step 17
From j1939-84 created by [mmschaefer](https://github.com/mmschaefer): SolidDesignNet/j1939-84#101  

Per:
[Task 5 clarification](https://github.com/SolidDesignNet/j1939-84/files/8505851/Task.5.clarification.1.pdf)

Add updates to requirements:
6.2.17.19	Actions10 for MY2022+ HEV and BEV drives
a. DS request message to ECU that indicated support in DM24 for upon request SP 12797 (Hybrid Lifetime Propulsion System Active Time) for 64241 PSA Times Lifetime Hours
b. Record each value for use in Part 12.
c. List data received in a table using lifetime, stored 100 hr, active 100hr for columns, and categories for rows.
6.2.17.20	Fail criteria10 for MY2022+ HEV and BEV drives (Where SP 12797 is supported in DM24)
a. Fail PG query where no response was received.  
b. Fail PG query where any accumulator value received is greater than FAFFFFFFh.
c. Fail all values where the corresponding value received in part 1 is greater than the part 2 value.
6.2.17.21	 Actions11 for MY2022+ HEV and BEV drives (Where SP 12797 is supported in DM24)
a. DS request message to ECU that indicated support in DM24 for upon request SP 12797 (Hybrid Lifetime Propulsion System Active Time) for Active 100hr PSA Times and Stored 100hr PSA Times PGs:
PG      PG Label                                  PG Acronym
64242 PSA Times Stored 100 Hours  PSATS 
64243 PSA Times Active 100 Hours   PSATA
b. Record each value for use in Part 12.
c. List data received in a table using lifetime, stored 100 hr, active 100hr for columns, and categories for rows.
6.2.17.22	 Fail criteria11 for MY2022+ HEV and BEV drives (Where SP 12797 is supported in DM24)
a. For MY2022+ HEV and BEV drives, Fail each PG query where no response was received.  
b. For MY2022-23 HEV and BEV drives, Warn each PG query, where no response was received.
c. Fail each PG query where any active technology label or accumulator value received is greater than FAFFh, respectively.
d. Fail all values where the corresponding value received in part 1 is greater than the part 2 value. (Where supported)
6.2.17.23	 Actions12 for MY2022+ Plug-in HEV DRIVES (Where SP 12783 is supported in DM24)
a. DS request message to ECU that indicated support in DM24 for upon request SP 12783 (Hybrid Lifetime Distance Traveled in Charge Depleting Operation with Engine off) for 64244 Hybrid Charge Depleting or Increasing Operation Lifetime Hours
b. Record each value for use in Part 12.
6.2.17.24	6.2.17.24 Fail criteria12 for MY2022+ Plug-in HEV DRIVES (Where SP 12783 is supported in DM24)
a. Fail PG query where no response was received.  
b. Fail PG query where any accumulator value received is greater than FAFFFFFFh.
c. Fail all values where the corresponding value received in part 1 is greater than the part 2 value.
6.2.17.25	Actions13 for MY2022+ Plug-in HEV DRIVES (Where SP 12783 is supported in DM24)
a. DS request message to ECU that indicated support in DM24 for upon request SP 12783 (Hybrid Lifetime Distance Traveled in Charge Depleting Operation with Engine off) for Active 100hr Charge Depleting times and Stored 100hr PSA Charge Depleting Times PGs:

PG       PG Label            PG Acronym
64246  Hybrid Charge Depleting or Increasing Operation Active 100 Hours  HCDIOA
64245  Hybrid Charge Depleting or Increasing Operation Stored 100 Hours   HCDIOS

b. Record each value for use in Part 12.
c. List data received in a table using lifetime, stored 100 hr, active 100hr for columns, and categories for rows.
6.2.17.26	Fail criteria13 for MY2022+ Plug-in HEV DRIVES (Where SP 12783 is supported in DM24)
a. For MY2024+ HEV and BEV drives, Fail each PG query where no response was received.  
b. For MY2022-23 HEV and BEV drives, Warn each PG query, where no response was received.
c. Fail each PG query where any active technology label or accumulator value received is greater than FAFFh, respectively.
d. Fail all values where the corresponding value received in part 1 is greater than the part 2 value. (Where supported)

### PR DESCRIPTION
**Pull request by mmschaefer**

### CODE REVIEW

File: src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java [hunk 1]
```diff
@@ -278,9 +309,592 @@ protected void run() throws Throwable {
...
+
+        }// end obdModule
+
+        // 6.2.17.6.g. Fail/warn per Table A-1, if two or more ECUs provide an SPN listed in Table A-1.
+        tableA1Validator.reportDuplicateSPNs(onRequestPackets,
+                                             getListener(),
+                                             "6.2.17.6.g");
+
+        // 6.2.17.6.h. Fail/warn per Table A-1 column, “Action if SPN provided but not included in DM24”
+        // @Joe we just need a method to call here. Eric added a column to Table A-1, I believe; please double check me
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java]: The method already exists.  It's `TableA1Validator.reportProvidedButNotSupportedSPNs()`

File: src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java [hunk 2]
```diff
@@ -278,9 +309,592 @@ protected void run() throws Throwable {
...
+                                                       GHG_TRACKING_LIFETIME_HYBRID_CHG_DEPLETING_PGs);
+            if (packetForPg == null) {
+                // 6.2.17.24.a. Fail PG query where no response was received
+                addFailure("6.2.17.24.a - No response was received from "
+                        + module.getModuleName() + "for PG "
+                        + GHG_TRACKING_LIFETIME_HYBRID_CHG_DEPLETING_PGs);
+            } else {
+                packetForPg.getSpns()
+                           .forEach(spn -> {
+                               // SAE INTERNATIONAL J1939TM-84 Proposed Draft 24 March
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java]: Please remove the cruft.

File: src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java [hunk 3]
```diff
@@ -278,9 +309,592 @@ protected void run() throws Throwable {
...
+                               // 2022 Page 37 of 140
+                               // 6.2.17.24.b - Fail PG query where any accumulator value
+                               // received is greater than FAFFFFFFh.
+                               if (spn.getRawValue() >= 0xFAFFFFFFL) {
+                                   addFailure("6.2.17.24.b - Bin value received is greater than 0xFAFFFFFF(h) from "
+                                           + module.getModuleName() + " for " + spn);
+                               }
+                               // FIXME: this needs to implemented once the datarepo is fixed
+                               // @Joe - just need to add the call once the dataRepo bug is fix
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java]: Is there a ticket filed for this bug?

**Review by gumbelmj — CHANGES_REQUESTED** [review]

File: src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java [hunk 4]
```diff
@@ -278,9 +309,593 @@ protected void run() throws Throwable {
...
+                    onRequestPackets.addAll(globalPackets);
                 }
+            }// end pgn
+
+            // 6.2.17.7 Actions4 for MY2022+ Diesel Engines
+            if (getEngineModelYear() >= 2022 && getFuelType().isCompressionIgnition()) {
+                getDataRepository()
+                                   .getObdModules()
+                                   .stream()
+                                   .filter(obdModuleInformation -> {
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java]: This filter is redundant with the checks below.

File: src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java (hunk 3 above)
Comment (mmschaefer) [reply src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java]: https://github.com/SolidDesignNet/j1939-84/issues/117

File: src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java [hunk 5]
```diff
@@ -278,9 +309,588 @@ protected void run() throws Throwable {
...
+
+        }// end obdModule
+
+        // 6.2.17.6.g. Fail/warn per Table A-1, if two or more ECUs provide an SPN listed in Table A-1.
+        tableA1Validator.reportDuplicateSPNs(onRequestPackets,
+                                             getListener(),
+                                             "6.2.17.6.g");
+
+        // 6.2.17.6.h. Fail/warn per Table A-1 column, “Action if SPN provided but not included in DM24”
+        onRequestPackets.stream()
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java]: `.stream()` is not necessary here.

File: src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java [hunk 6]
```diff
@@ -278,9 +309,588 @@ protected void run() throws Throwable {
...
+                               if (spn.getRawValue() >= 0xFAFFL) {
+                                         addFailure("6.2.17.26.c - Bin value received is greater than 0xFAFF(h) from "
+                                                 + module.getModuleName() + " for " + spn);
+                               }
+                               // FIXME: this needs to be implemented once the dataRepo is fixed
+                               // @Joe: just need to add the call and warning when the dataRepo bug is fixed.
+                               /// 6.2.17.26.d - Fail each PG query where any active
+                               /// technology label or accumulator value
+                               // received is greater than FAFFh, respectively.
+                               // if (spn.getRawValue() >= 0xFAFFL) {
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java]: Why is this code commented out?

File: src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java [hunk 7]
```diff
@@ -278,9 +309,588 @@ protected void run() throws Throwable {
...
+                packetForPg.getSpns()
+                           .forEach(spn -> {
+                               // 6.2.17.20.b - Fail PG query where any accumulator value
+                               // received is greater than FAFFFFFFh.
+                               if (spn.getRawValue() >= 0xFAFFFFFFL) {
+                                   addFailure("6.1.26.20.b - Bin value received is greater than 0xFAFFFFFF(h) from "
+                                           + module.getModuleName() + " for " + spn);
+                               }
+                               // FIXME: need to add functionality when dataRepo is updated
+                               // 6.2.17.20.c - Fail all values where the corresponding values received
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java]: The rest of the text is missing here.

File: src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java [hunk 8]
```diff
@@ -278,9 +309,588 @@ protected void run() throws Throwable {
...
+                    addWarning("6.2.17.14.b - No response was received from "
+                            + module.getModuleName() + " for PG "
+                            + pg);
+                }
+            } else {
+                trackingPacketForPg.getSpns()
+                           .forEach(spn -> {
+                               if (spn.getRawValue() > 0xFAFFL) {
+                                   // 6.2.17.14.c - Fail each PG query where any value received is greater than FAFFh.
+                                           addFailure("6.1.26.14.c - Bin value received is greater than 0xFAFF from "
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java]: Section number is wrong here.

File: src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java [hunk 9]
```diff
@@ -278,9 +309,588 @@ protected void run() throws Throwable {
...
+                packetForPg.getSpns()
+                           .forEach(spn -> {
+                               // 6.2.17.18.c. Fail PG query where any bin value received is greater than FAFFh.
+                               if (spn.getRawValue() >= 0xFAFFL) {
+                                   addFailure("6.2.17.18.c - Bin value received is greater than 0xFAFF(h) from "
+                                           + module.getModuleName() + " for " + spn);
+                               }
+                               // 6.2.17.18.d. Fail PG query where any index value received is greater than FAh.
+                               if (spn.getId() == 12691 && spn.getValue() > 0xFAL) {
+                                   addFailure("6.2.17.18.c - PG query index received was " + "");
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java]: Section number is wrong here.

File: src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java [hunk 10]
```diff
@@ -278,9 +309,588 @@ protected void run() throws Throwable {
...
+                        // FAFFFFFFh for NOx values)
+                        addFailure("6.2.17.10.c - Bin value received is greater than 0xFAFFFFFF(h) from "
+                                + module.getModuleName() + " for " + spn);
+
+                    }
+                    // 6.2.17.10.e. Fail each active 100 hr array value that is greater than zero. (where supported)
+                    if (spn.getValue() > 0) {
+                        // 6.2.17.10.d. Warn for all active 100 hr bin 3 through bin 17 values that are greater than
+                        // zero. (Where supported)
+                        addWarning("6.2.17.10.d - Active 100 hr array bin 3 through bin 17 values received has a value greater than zero (where supported) from "
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java]: Please condense this text.

**Review by gumbelmj — CHANGES_REQUESTED** [review]

File: src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java [hunk 11]
```diff
@@ -278,9 +309,571 @@ protected void run() throws Throwable {
...
+            addWarning("6.2.17.20.a - No response was received from "
+                    + module.getModuleName() + " for PG "
+                    + GHG_TRACKING_LIFETIME_HYBRID_PGs);
+        } else {
+            packetForPg.getSpns()
+                       .forEach(spn -> {
+                           // 6.2.17.20.b - Fail PG query where any accumulator value
+                           // received is greater than FAFFFFFFh.
+                           if (spn.getRawValue() >= 0xFAFFFFFFL) {
+                               addFailure("6.1.26.20.b - Bin value received is greater than 0xFAFFFFFF(h) from "
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java]: Section number is wrong.

File: src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java [hunk 12]
```diff
@@ -278,9 +309,571 @@ protected void run() throws Throwable {
...
+                packetForPg.getSpns()
+                           .forEach(spn -> {
+                               // 6.2.17.18.c. Fail PG query where any bin value received is greater than FAFFh.
+                               if (spn.getRawValue() >= 0xFAFFL) {
+                                   addFailure("6.2.17.18.c - Bin value received is greater than 0xFAFF(h) from "
+                                           + module.getModuleName() + " for " + spn);
+                               }
+                               // 6.2.17.18.d. Fail PG query where any index value received is greater than FAh.
+                               if (spn.getId() == 12691 && spn.getValue() > 0xFAL) {
+                                   addFailure("6.2.17.18.d - PG query index received was " + "");
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/controllers/part02/Part02Step17Controller.java]: The rest of this failure message is missing.

**Review by gumbelmj — APPROVED** [review]
There are two minor changes that need resolved and then you can merge it.

**Pull request merged**